### ISSUE TITLE
Not directly related to oxipng, but do you or anyone else know of something that works similarly for .ico files?

### ISSUE THREAD

**Issue by Jieiku** [shssoichiro/oxipng#440]
I absolutely love oxipng, it is my new favorite PNG optimizer.

I have been using it like this (Is it possible to get more compression? did I miss any of the advanced options?):

`oxipng -o max --strip all -a -Z *.png`

I am wondering if there is a tool that does something similar for .ico files? (I still have one of these for old browser compatibility)

**Comment by TPS** [comment]
[FileOptimizer](https://github.com/shssoichiro/oxipng/issues/326#issuecomment-711372708) has a toolchain for `.ico`.

**Comment by Jieiku** [comment]
Seems to be a windows product (unless i am wrong?) If anyone know of something that runs on Linux let me know, preferably with the command line but does not have to be.

I can try running this under wine if I get some time to mess around with that.

**Comment by TPS** [comment]
According to [FO's current code](https://sourceforge.net/p/nikkhokkho/code/HEAD/tree/trunk/FileOptimizer/Source/cppMain.cpp#l1526], it runs (directly on .ico) in this order

- magick.exe convert `INPUTFILE` -quiet -compress ZIP

- [leanify.exe](https://github.com/JayXon/Leanify) -q -p

Probably tweak those parameters some, as FO changes them based on GUI options.  Both tools have Linux releases.

**Comment by Jieiku** [comment]
ahhh good eye, I gave that a go, but does not work:
![2022-05-29_12-37-00](https://user-images.githubusercontent.com/106644/170888485-8bb25833-9406-4837-8db9-fc33ea0002e4.png)

**Comment by TPS** [comment]
leanify v0.4.3 is ancient.  Have you tried building from current source?  Did you try running imagemagick w/ those parameters 1st?  I don't hear much from anyone re: optimizing `.ico` & I've had mixed results myself.

**Comment by TPS** [comment]
He hasn't released binaries in very long time.  You have to build from source to get 7 years of improvements.  Last commit 4 months ago even improved Linux compilation.

**Comment by TPS** [comment]
Any results?

**Comment by Jieiku** [comment]
Had something come up with a short deadline, going to take another look at this tomorrow.

**Comment by Jieiku** [comment]
Gave this a try this evening, but was unable to build, I opened a ticket there though: https://github.com/JayXon/Leanify/issues/84

**Comment by TPS** [comment]
No 1 tool is perfect, or even close to it by itself, but using many can approach state-of-the-art.  Try FileOptimizer any way you can — it's the best I've found in the last decade or so.

**Comment by javiergutierrezchamorro** [comment]
> [FileOptimizer](https://github.com/shssoichiro/oxipng/issues/326#issuecomment-711372708) has a toolchain for `.ico`.

Thank you for suggesting

**Comment by TPS** [comment]
> Thank you for suggesting

@javiergutierrezchamorro  Thank _you_ for continuing FO all these years.

**Comment by Zerowalker** [comment]
Would be nice if Oxipng added support for .ico, though I get that it's niche.
But isn't ICO essentially the same as a png file, just a different structure to support multiple files in one?

**Comment by TPS** [comment]
> But isn't ICO essentially the same as a png file, just a different structure to support multiple files in one?

Umm, no.  [`.ico` can _contain_ `.png`, but is otherwise entirely a different format.](https://en.wikipedia.org/wiki/ICO_(file_format))

**Comment by TPS** [comment]
> Would be nice if Oxipng added support for .ico, though I get that it's niche.

You're essentially looking for [`leanify`](https://github.com/shssoichiro/oxipng/issues/440#issuecomment-1150908340), which is designed to _also_ work on archive formats, which `.ico` essentially is.

**Comment by andrews05** [comment]
Maybe you could suggest leanify adopt oxipng 😄
FileOptimizer does include it now.

**Comment by Zerowalker** [comment]
> > But isn't ICO essentially the same as a png file, just a different structure to support multiple files in one?
> 
> Umm, no. [`.ico` can _contain_ `.png`, but is otherwise entirely a different format.](https://en.wikipedia.org/wiki/ICO_(file_format))

That's basically what I meant, but thought it was only png files, so my bad:)

**Comment by dustdfg** [comment]
Does anyone knows similar thing but for webp?

**Comment by TPS** [comment]
> [FileOptimizer](https://github.com/shssoichiro/oxipng/issues/440#issuecomment-1140345986) might be even better.

@dustdfg In general, this is the answer to all such questions. @javiergutierrezchamorro is _very_ diligent in collecting optimizers in FO.

**Comment by javiergutierrezchamorro** [comment]
> > [FileOptimizer](https://github.com/shssoichiro/oxipng/issues/440#issuecomment-1140345986) might be even better.
> 
> @dustdfg In general, this is the answer to all such questions. @javiergutierrezchamorro is _very_ diligent in collecting optimizers in FO.

Thank you!

**Comment by dustdfg** [comment]
> > [FileOptimizer](https://github.com/shssoichiro/oxipng/issues/440#issuecomment-1140345986) might be even better.
> 
> @dustdfg In general, this is the answer to all such questions. @javiergutierrezchamorro is _very_ diligent in collecting optimizers in FO.

Yeah, I hate windows and switched from that crap years ago and I think I don't need it that much if it requires windows experience again but thank you for recommendation

**Comment by TPS** [comment]
> Yeah, I hate windows and switched from that crap years ago and I think I don't need it that much if it requires windows experience again but thank you for recommendation

@dustdfg FO runs fine under Wine.  Try & see.

@javiergutierrezchamorro As always, thank _you_!

**Comment by goodusername123** [comment]
A hidden gem that exists on GitHub is [IcoTools](https://github.com/jtippet/IcoTools/) which provides a full suite of CLI tools for doing things with Windows ico files including being able to use external PNG optimizers such as oxipng to optimize ico files as well!

example(s) of using IcoCrush from [IcoTools](https://github.com/jtippet/IcoTools/) with oxipng:

- `IcoCrush -i "Application-Icon.ico" --write-in-place --preferred-format PngLargeImages --png-tool "oxipng -v -o max -Z -a --nb --nc"` (generates icons compatible with versions of Windows older then Vista)
- `IcoCrush -i "Application-Icon.ico" --write-in-place --preferred-format MinimizeStorage --png-tool "oxipng -v -o max -Z -a --nb --nc"` (sacrifices compatibility with versions of Windows older then Vista for maximum compression)
